### ISSUE TITLE
[feature request] Callback handler event after forward pass in Trainer

### ISSUE THREAD

**Issue by yanadrdr** [huggingface/transformers#36012]
While working on training (LORA) some model, I wanted to track memory usage in reference with meaningful events: 
1. after forward pass - to track activation memory consumption
2. after forward pass - to track gradient and optimizer memory consumption
3. after zero grad - to make sure everything is released and we go back to only model weights consuming memory.

It was very easy to set up 2 and 3, however I noticed that there is no event related to the forward pass. To inject the wanted behavior I had to override the compute loss although loss computation stayed exactly the same. I think an event related to forward pass will greatly reduce the need to override compute_loss to inject custom behavior (that is not related to loss computation), and will enable cleaner more robust code by allowing to just use callback in such cases.

It might not be relevant to the feature request, but here is the callback I used to log memory metrics after backward and zero grad:
```
class MemoryStatsCallback(TrainerCallback):
    def __init__(self, filepath, step_log_interval):
        super().__init__()
        self.filepath = filepath
        os.makedirs(os.path.dirname(self.filepath), exist_ok=True)
        if os.path.exists(self.filepath):
            os.remove(self.filepath)
        self.global_step = -1
        self.step_log_interval = step_log_interval
        self.log_flag = False

    def _increment_step_counter(self):
        self.global_step += 1
        if self.global_step % self.step_log_interval == 0:
            self.log_flag = True
        else:
            self.log_flag = False

    def write_memory_stats_to_file(self, event: str):
        if not self.log_flag:
            return
        allocated_memory = get_used_gpu_memory()
        with open(self.filepath, 'a') as f:
            f.write(f'{event} {allocated_memory}\n')

    def on_step_begin(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        self._increment_step_counter()
        return control

    def on_substep_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        # substep is forward-backward during gradient accumulation (if not finished the grad accum steps)
        self.write_memory_stats_to_file(event='backward')
        return control

    def on_pre_optimizer_step(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        # after backward before doing optimizer.step
        self.write_memory_stats_to_file(event='backward')
        return control

    def on_step_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        # after zero grad
        self.write_memory_stats_to_file(event='zero_grad')
        torch.cuda.empty_cache()

        return control
```
To also account for memory metrics after forward pass I overriden the compute_loss

**Comment by Rocketknight1** [comment]
cc @muellerzr @SunMarc

**Comment by SunMarc** [comment]
I think that for this kind of details, I recommend you using a profiler instead. It will give you all the details you want. Have a look at this blogpost https://pytorch.org/blog/understanding-gpu-memory-1/

**Comment by yanadrdr** [comment]
Thanks, actually looked at it before resorting to the memory callback but it was a bit too much and did not give me the explicit report I wanted (for example if the activations in the forward pass do not take up much space I guess it will be hard to understand exactly where is the forward and where is the backward memory stats, that's why I wanted something explicit)

**Comment by SunMarc** [comment]
Thanks for the context ! Yeah, we can add callback after the forward pass of the model during training. Would you like to open a PR to add it ?

**Comment by yanadrdr** [comment]
Yes, thanks! 
It is my first contribution so might take a while, will go over the guidelines and start working on it

**Comment by yanadrdr** [comment]
Hi, not sure how to proceed, prepared a PR and passed reviews but I am failing on tests that look unrelated to the changes I have made. Tried the "update branch" a few times but doesn't seem to make a difference. How should I debug/fix this?
https://github.com/huggingface/transformers/pull/36496
Thanks!

**Comment by SunMarc** [comment]
> Hi, not sure how to proceed, prepared a PR and passed reviews but I am failing on tests that look unrelated to the changes I have made. Tried the "update branch" a few times but doesn't seem to make a difference. How should I debug/fix this?
> https://github.com/huggingface/transformers/pull/36496
> Thanks!

This is not your fault. Our CI is a bit flaky these days. I will keep an eye on it and merge it when the tests are passing.

**Comment by GauravBh1010tt** [comment]
Hey,
I need this functionality so I can experiment with the custom loss. I am wondering if there is any update on this? 
Thanks
